### PR TITLE
 HPCC-19925 Add not present in the environment indicator

### PR DESCRIPTION
**Pull request by kenrowland**
Add flag indicating that an attribute is not present in the environment

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by kenrowland** [comment]
@rpastrana Please review. Only concentrate on the 2nd commit. The first commit is the changes for another PR that is not merged, but required for this PR. When approved will be rebased to master after dependent PR is merged.

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (rpastrana) [inline esp/services/ws_config2/ws_config2Service.cpp]: attributeName is only used once, why not pAttribute->setName(pSchemaValue->getName().c_str());

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (rpastrana) [inline esp/services/ws_config2/ws_config2Service.cpp]: why not use pLimits

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (rpastrana) [inline esp/services/ws_config2/ws_config2Service.cpp]: pLimits?

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (rpastrana) [inline esp/services/ws_config2/ws_config2Service.cpp]: pLimits?

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (rpastrana) [inline esp/services/ws_config2/ws_config2Service.cpp]: pLimits?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@kenrowland a few comments

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (kenrowland) [reply esp/services/ws_config2/ws_config2Service.cpp]: An environment value is created with a default schema value type as well. Essentially it is treated as a basic string type with no value verification. The user can make changes to the value and it will be written to the environment when saved.

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (kenrowland) [reply esp/services/ws_config2/ws_config2Service.cpp]: There is no particular reason other than getName returns a const reference and I wanted to avoid possible compile problems down the road should attributeName be needed for another purpose. I went ahead and changed it.

File: esp/services/ws_config2/ws_config2Service.cpp
Comment (kenrowland) [reply esp/services/ws_config2/ws_config2Service.cpp]: Changed here and other places as well.

**Comment by kenrowland** [comment]
@rpastrana Please review latest commit based on comments.

**Review by rpastrana — APPROVED** [review]
@kenrowland

**Comment by kenrowland** [comment]
@richardkchapman please merge

**Comment by kenrowland** [comment]
@ghalliday Sorry, can you please merge this one as well. Thanks

File: esp/scm/ws_config2.ecm
Comment (ghalliday) [inline esp/scm/ws_config2.ecm]: Should this have a default value?

**Comment by ghalliday** [comment]
@kenrowland a couple of minor questions.

File: esp/scm/ws_config2.ecm
Comment (kenrowland) [reply esp/scm/ws_config2.ecm]: It is always assigned a value in the code.

**Pull request merged**
